core/tests/test_images.py

```diff
@@ -39,87 +45,158 @@ def test_create_from_file(app, args, kwargs):
     ):
         toga.Image(None)
 
 
-def test_too_many_arguments():
+def test_invalid_input_format():
+    """Trying to create an image with an invalid input should raise an error"""
     with pytest.raises(
-        ValueError,
-        match=r"Only either path or data can be set.",
+        TypeError,
+        match=r"Unsupported source type for Image",
     ):
-        toga.Image(path="/image.png", data=bytes([1]))
+        toga.Image(42)
+
+
+def test_create_from_pil(app):
+    """An image can be created from a PIL image"""
+    pil_img = PIL.Image.open(ABSOLUTE_FILE_PATH)
+    toga_img = toga.Image(pil_img)
+
+    assert isinstance(toga_img, toga.Image)
 
 
-def test_dimensions():
-    "The dimensions of the image can be retrieved"
+@pytest.mark.parametrize("kwargs", [{"data": BYTES}, {"path": ABSOLUTE_FILE_PATH}])
+def test_deprecated_arguments(kwargs):
+    with pytest.deprecated_call():
+        toga.Image(**kwargs)
 
-    image = toga.Image(path="resources/toga.png")
+
+@pytest.mark.filterwarnings("ignore::DeprecationWarning")
+@pytest.mark.parametrize(
+    "args, kwargs",
+    [
+        # One positional, one keyword
+        ((ABSOLUTE_FILE_PATH,), {"data": BYTES}),
+        ((ABSOLUTE_FILE_PATH,), {"path": ABSOLUTE_FILE_PATH}),
+        # Two keywords
+        ((), {"data": BYTES, "path": ABSOLUTE_FILE_PATH}),
+        # All three
+        ((ABSOLUTE_FILE_PATH,), {"data": BYTES, "path": ABSOLUTE_FILE_PATH}),
+    ],
+)
+def test_too_many_arguments(args, kwargs):
+    """If multiple arguments are supplied, an error is raised"""
+    with pytest.raises(
+        ValueError,
+        match=r"Received multiple arguments to constructor.",
+    ):
+        toga.Image(*args, **kwargs)
+
+
+def test_dimensions(app):
+    """The dimensions of the image can be retrieved"""
+    image = toga.Image(RELATIVE_FILE_PATH)
 
     assert image.size == (60, 40)
     assert image.width == 60
     assert image.height == 40
 
 
-def test_data():
-    "The raw data of the image can be retrieved."
-    image = toga.Image(path="resources/toga.png")
-
+def test_data(app):
+    """The raw data of the image can be retrieved."""
+    image = toga.Image(ABSOLUTE_FILE_PATH)
     assert image.data == b"pretend this is PNG image data"
 
 
-def test_image_save():
-    "An image can be saved"
-    save_path = Path("/path/to/save.png")
-    image = toga.Image(path=ABSOLUTE_FILE_PATH)
-
+def test_image_save(tmp_path):
+    """An image can be saved"""
+    save_path = tmp_path / "save.png"
+    image = toga.Image(BYTES)
     image.save(save_path)
+
     assert_action_performed_with(image, "save", path=save_path)
+
+
+def test_as_format_pil(app):
+    """as_format can successfully return a PIL image"""
+    toga_img = toga.Image(ABSOLUTE_FILE_PATH)
+    # The dummy backend only supplies toy data, but this verifies that it is being
+    # fed to PIL.Image.open
+    with pytest.raises(PIL.UnidentifiedImageError):
+        toga_img.as_format(PIL.Image.Image)
```

Comment:
Hrm... the limitation makes sense given the current dummy data; however, given this test case, I think it might be worth fixing the backend to use "real" data - otherwise the implementation of as_format could be completely broken, and as long as it passes *something* to the PIL constructor, it would pass this test.

Maybe we could modify dummy.Image so it's a *little* smarter - if the path actually exists on disk, it reads the file and returns it as bytes (and only returns the dummy data if a fake path is provided)? Then, use on of the toga resource images as the actual data, and validate we get a PIL 32x32 PNG image?

Reply:
I actually started to go down this route at first, but looking at the rest of the dummy implementation, I worried I was deviating from what it was intended to do. I've now implemented this as sensibly as I could, which required changes to the Window and Canvas dummies and tests as well.